govcd/user.go

```diff
@@ -0,0 +1,463 @@
+			user.User.GroupReferences = nil
+		}
+	}
+
+	userHREF, err := url.ParseRequestURI(user.User.Href)
+	if err != nil {
+		return fmt.Errorf("error getting HREF for user %s : %v", user.User.Name, err)
+	}
+	util.Logger.Printf("[TRACE] Url for updating user : %#v and name: %s", userHREF, user.User.Name)
+
+	_, err = user.client.ExecuteRequest(userHREF.String(), http.MethodPut,
+		types.MimeAdminUser, "error updating user : %s", user.User, nil)
+	return err
+}
+
+// Disable disables a user, if it is enabled. Fails otherwise.
+func (user *OrgUser) Disable() error {
+	util.Logger.Printf("[TRACE] Disabling user: %s", user.User.Name)
+
+	if !user.User.IsEnabled {
+		return fmt.Errorf("user %s is already disabled", user.User.Name)
+	}
+	user.User.IsEnabled = false
+
+	return user.Update()
+}
+
+// ChangePassword changes user's password
+// Constraints: the password must be non-empty, with a minimum of 6 characters
+func (user *OrgUser) ChangePassword(newPass string) error {
+	util.Logger.Printf("[TRACE] Changing user's password user: %s", user.User.Name)
+
+	user.User.Password = newPass
+
+	return user.Update()
+}
+
+// Enable enables a user if it was disabled. Fails otherwise.
+func (user *OrgUser) Enable() error {
+	util.Logger.Printf("[TRACE] Enabling user: %s", user.User.Name)
+
+	if user.User.IsEnabled {
+		return fmt.Errorf("user %s is already enabled", user.User.Name)
+	}
+	user.User.IsEnabled = true
+
+	return user.Update()
+}
+
+// Unlock unlocks a user that was locked out by the system.
+// Note that there is no procedure to LOCK a user: it is locked by the system when it exceeds the number of
+// unauthorized access attempts
+func (user *OrgUser) Unlock() error {
+	util.Logger.Printf("[TRACE] Unlocking user: %s", user.User.Name)
+
+	if !user.User.IsLocked {
+		return fmt.Errorf("user %s is not locked", user.User.Name)
+	}
+	user.User.IsLocked = false
+
+	return user.Update()
+}
+
+// ChangeRole changes a user's role
+// Fails is we try to set the same role as the current one.
+// Also fails if the provided role name is not found.
+func (user *OrgUser) ChangeRole(roleName string) error {
+	util.Logger.Printf("[TRACE] Changing user's role: %s", user.User.Name)
+
+	if roleName == "" {
+		return fmt.Errorf("role name cannot be empty")
+	}
+
+	if user.User.Role != nil && user.User.Role.Name == roleName {
+		return fmt.Errorf("new role is the same as current role")
+	}
+
+	newRole, err := user.adminOrg.GetRole(roleName)
+	if err != nil {
+		return err
+	}
+	user.User.Role = newRole
+
+	return user.Update()
+}
+
+// TakeOwnership takes ownership of the user's objects.
+// Ownership is transferred to the caller.
+// This is a call to make before deleting. Calling user.SafeDelete() will
+// run TakeOwnership before the actual user removal.
+// API Documentation: https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-TakeOwnership.html
+func (user *OrgUser) TakeOwnership() error {
```

Comment:
Nice approach. I'm just wondering - is there any risk of "canibalizing" the objects? I mean if I'm running as a `sysadmin` on some org and then while deleting it sysadmin takes ownership. Does it then mean that no one in that specific org is able to perform actions on it except the `sysadmin`?

Reply:
The approach is crude, but we don't have much choice.
If we need to delete a user, we can :
a) take ownership of the objects and then deal with privileges by reassigning ownership appropriately;
b) stop any running vApp/VM belonging to the user and then delete the user. This will also delete all vApps.

I assume that deleting a user is always a drastic decision, and there are going to be consequences. Scenario "a" seems the one that gives time to handle things nicely, while scenario "b" seems appropriate in case of a security threat.

Reply:
What todo if you want delete User with all dependencies?

Reply:
you can do it only if the dependencies are not running vApps/VMs. If they are stopped, and you don't run take ownership first, the dependencies will be deleted together with the user

Reply:
Do we have capability to find dependent vApps/VMs ?

Reply:
> Do we have capability to find dependent vApps/VMs ?

With our code, not starting from the user. We can follow the `Owner` structure of every vApp/VM and see to whom they belong.

Reply:
For the first iteration, I'd keep it simple. Just try to delete the user and shine-through any resulting API error (assuming it says that it can't be deleted because of that and that).

Reply:
That's what it does

Reply:
After reading this after a break, I realize that the real issue here is not with `TakeOwnership()`, it's with naming of `SafeDelete()` which calls it.

The word "Safe" here is counter-intuitive, because actually it's on the opposite of safe - it will delete the user even if there are objects belonging to it. It's better to rename `SafeDelete()` to something like:
- `DeleteForce()`
- `DeleteTakeOwnership()`

p.s. As in these suggestions, we should also use the CRUD naming - "Delete" being at the beginning of the method.

CC: @Didainius @vbauzysvmware

Reply:
Agree about beginning with *Delete*

Reply:
Me too